List timestamps

endophage opened the pull request:
Server code for listing timestamps.

It doesn't implement a rigid pagination but instead relies on clients to keep track of pages themselves, letting them send a non-inclusive starting hash and a number of items to return.

endophage commented:
Moving back to WIP. After discussions we've decided the API should be <checksum> and <number of entries> as the inputs, not a start index (the server will loop up the start index in the DB using the checksum).

cyli commented on server/handlers/default.go:
```diff
@@ -211,6 +214,65 @@ func getKeyHandler(ctx context.Context, w http.ResponseWriter, r *http.Request,
+	return
+}
+
+func listVersionsHandler(ctx context.Context, w http.ResponseWriter, store storage.MetaStore, gun, role, start string, number int) error {
```
Non-blocking - I think in a previous PR, you were going to change these handler functions to just take a writer of any kind, which seemed like a pretty cool idea.  :D  Is that possible here?  This function doesn't seem to use the context, for instance.

cyli commented on server/handlers/default.go:
```diff
@@ -211,6 +214,65 @@ func getKeyHandler(ctx context.Context, w http.ResponseWriter, r *http.Request,
+}
+
+func parsePageParams(f url.Values) (start string, number int, err error) {
+	start = f.Get("start")
```
Non-blocking - for an HTTP API, unless we are trying to make this pagination more general (so it applies to more things), maybe more descriptive query parameters would make it easier to understand how it works?  Maybe something like `from_checksum` and `num_versions` or some such?  (I'm bad at naming, so probably something else :D)

cyli replied:
Update: or maybe `before_checksum` or the like, since the results seem to not include the starting checksum?

cyli commented on server/handlers/default.go:
```diff
@@ -211,6 +214,65 @@ func getKeyHandler(ctx context.Context, w http.ResponseWriter, r *http.Request,
+}
+
+func listVersionsHandler(ctx context.Context, w http.ResponseWriter, store storage.MetaStore, gun, role, start string, number int) error {
+	versions, err := store.GetVersions(gun, role, start, number)
```
If `0` is a valid number of results to return, should we add an explicit check for it here and bail early, rather than make a DB query?

cyli replied:
Oh oops, sorry, I missed the part where `0` meant "get all".  nm!  :)

cyli commented on server/errors/errors.go:
```diff
@@ -26,6 +26,12 @@ var (
+	ErrBadPagination = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "BAD_PAGINATION",
+		Message:        "The pagination parameters are invalid.",
+		Description:    "Either the start or number parameters could not be parsed as integers greater than or equal to zero.",
```
Maybe "The number parameter could not be parsed as an integer greater...", since we don't check the start except to see if it exists?  (If it doesn't, we return `ErrNotFound`, and not `ErrBadPagination`)

cyli commented on server/storage/database_test.go:
```diff
@@ -501,3 +482,133 @@ func TestDBGetChecksumNotFound(t *testing.T) {
+		MetaUpdate{Role: "timestamp", Version: 1, Data: tsJSON1},
+	)
+
+	// table is empty
```
Nitpick:  I think this comment might have been a copy-paste artifact?

cyli commented on server/storage/database_test.go:
```diff
@@ -501,3 +482,133 @@ func TestDBGetChecksumNotFound(t *testing.T) {
+	require.EqualValues(t, tsJSON1, versions[0])
+}
+
+func TestGetVersionsNotFound(t *testing.T) {
```
Awesome test with all the cases of DB failure!  :D

cyli commented on server/storage/database_test.go:
```diff
@@ -501,3 +482,133 @@ func TestDBGetChecksumNotFound(t *testing.T) {
+	require.IsType(t, ErrNotFound{}, err)
+}
+
+func TestTranslateOldVersionError(t *testing.T) {
```
Yay!  Thanks for adding this!

endophage replied:
It helped me get the test coverage to acceptable levels :-P

cyli commented on server/storage/interface.go:
```diff
@@ -34,6 +34,18 @@ type MetaStore interface {
+	// generated/validated from this raw json by the consumer.
+	// start is a checksum which will be used as a non-inclusive offset
+	// An empty start value means start from (and include) the most recent version
+	// numToResturn can limit how many results are returned.
```
Nitpick: "numToReturn"

cyli commented on server/storage/database_test.go:
```diff
@@ -501,3 +482,133 @@ func TestDBGetChecksumNotFound(t *testing.T) {
+	require.NoError(t, err)
+
+	require.Len(t, versions, 1)
+	require.EqualValues(t, tsJSON1, versions[0])
```
Would it be possible to add a couple more checks here?  Some of these may be redundant (so please feel free to reject any or all):

1. `store.GetVersions("gun", "timestamp", "", 1)` should return just `versions[2]` (I think?)
1. `store.GetVersions("gun", "timestamp", checksum, 0)` should return `versions[0]`
1. `store.GetVersions("gun", "timestamp", checksum, 80)` should also return `versions[0]`
1. `store.GetVersions("gun", "timestamp", "", 80)` should also return all?

The number being > than what's there case is the particular one I wanted to test.

endophage replied:
I totally missed testing what happens if the limit is larger than the total number of possible entries to return. Great catch.

cyli commented on server/storage/memory.go:
```diff
@@ -84,11 +90,50 @@ func (st *MemStorage) GetCurrent(gun, role string) (data []byte, err error) {
+	)
+	if start != "" {
+		for st.tufMeta[id][idx].checksum != start {
+			if idx == len(st.tufMeta[id]) {
```
Should this be `len(st.tufMeta[id]) - 1`?  By the time it reaches `len(st.tufMeta[id])`, won't this be an index out of range error?  Perhaps we could add a test for searching an in-memory store for a checksum that doesn't exist?

cyli commented:
Could we also add a test in `server/handlers/default_test.go`'s `TestListVersionsHandler` that passing an invalid number (not parseable as string or negative number) in returns the predicted error? 

Also, can we add a test in `server/server_test.go`'s `TestListVersions` that a URL route that includes the start but no number returns an error (is it a 404?)

endophage commented:
@cyli renamed `start` to `cursor` and `number` to `limit` (I genreally prefer generic terms so that they can be used consistently throughout an HTTP API rather than having custom parameter names for every endpoint).

A URL with a start but no number wouldn't (and shouldn't) have been an error, it just wouldn't have had the regex checks applied to ensure the checksum was valid. It would have been caught by the URL route for versions that had no matching query params. That made me realize any request to versions for which the params didn't match the regexes would get caught by that version of the route so I've just removed the other ones and added a check to parse params to make sure the cursor is a valid sha256 hex checksum.

riyazdf commented on server/handlers/default.go:
```diff
@@ -211,6 +216,76 @@ func getKeyHandler(ctx context.Context, w http.ResponseWriter, r *http.Request,
+	numberStr := f.Get("limit")
+	if numberStr != "" {
+		if number, err = strconv.Atoi(numberStr); err != nil || number < 0 {
+			return "", 0, fmt.Errorf("could not parse number parameter")
```
should this read "limit" instead of "start"?

endophage replied:
yep, fixed to "limit"

diogomonica commented:
Closing for now.

endophage commented:
The request for this feature has been tabled so this would have been unused code for some period. We will revisit this and either rewrite, or re-open as and when appropriate.
